source/Halibut/Transport/TcpConnectionFactory.cs

```diff
@@ -52,16 +54,19 @@ public async Task<IConnection> EstablishNewConnectionAsync(ExchangeProtocolBuild
             var certificateValidator = new ClientCertificateValidator(serviceEndpoint);
             var client = await CreateConnectedTcpClientAsync(serviceEndpoint, log, cancellationToken);
             log.Write(EventType.Diagnostic, $"Connection established to {client.Client.RemoteEndPoint} for {serviceEndpoint.BaseUri}");
-
-            var stream = client.GetStream();
+            
+            var networkStream = client.GetStream();
+            var networkTimeoutStream = new NetworkTimeoutStream(networkStream);
+            var ssl = new SslStream(networkTimeoutStream, false, certificateValidator.Validate, UserCertificateSelectionCallback);
 
             log.Write(EventType.SecurityNegotiation, "Performing TLS handshake");
-            var ssl = new SslStream(stream, false, certificateValidator.Validate, UserCertificateSelectionCallback);
-            
-            // TODO - ASYNC ME UP!
-            // This should take a cancellation token.
```

Comment:
Hey @LukeButters, I was just reflecting on this comment.

Technically, we didn't do this for .NET 4.8.

.NET 6.0 will do it, as we are calling the override with the cancellation token.

For .NET 4.8, we now respect the timeouts of the underlying steam. So it's an improvement. But there is no cancellation token being passed into `AuthenticateAsClientAsync` (so nowhere to hook into).

The consequence is that if we cancel this operation, then `AuthenticateAsClientAsync` will ignore that. A cancellation exception will get thrown on the very next line though. So probably not a big deal.

If we were going to handle this, maybe we'd have to pass the cancellation token into the constructor of `NetworkTimeoutStream`, which it can use for all its internal methods. This is a bit yuck, as it means we lock in the cancellation token of the calling method, even if the stream is returned and used outside. But it would work in this case.

I'd lean towards leaving it as is. But wanted to see what you thought.

Reply:
Just mark it as async me up and add a task to the main convert ticket. Don't forget to add a comment around net48 does not support cancellation tokens here.

Reply:
Done